### ISSUE TITLE
problem with upgrading HA masters from v1.2 to v1.4

### ISSUE THREAD

**Issue by DreadPirateShawn** [kubernetes/kubernetes#34265]
**Kubernetes version**

Upgrading from:

```
Client Version: version.Info{Major:"1", Minor:"2", GitVersion:"v1.2.2", GitCommit:"528f879e7d3790ea4287687ef0ab3f2a01cc2718", GitTreeState:"clean"}
Server Version: version.Info{Major:"1", Minor:"2", GitVersion:"v1.2.2", GitCommit:"528f879e7d3790ea4287687ef0ab3f2a01cc2718", GitTreeState:"clean"}
```

to

```
Client Version: version.Info{Major:"1", Minor:"4", GitVersion:"v1.4.0", GitCommit:"a16c0a7f71a6f93c7e0f222d961f4675cd97a46b", GitTreeState:"clean", BuildDate:"2016-09-26T18:16:57Z", GoVersion:"go1.6.3", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"4", GitVersion:"v1.4.0", GitCommit:"a16c0a7f71a6f93c7e0f222d961f4675cd97a46b", GitTreeState:"clean", BuildDate:"2016-09-26T18:10:32Z", GoVersion:"go1.6.3", Compiler:"gc", Platform:"linux/amd64"}
```

**Environment**:
- **Cloud provider or hardware configuration**: Bare metal
- **OS** (e.g. from /etc/os-release): Ubuntu 12.04.4 LTS, Precise Pangolin
- **Kernel** (e.g. `uname -a`): Linux tuk6r1knode3 3.13.0-55-generic #92~precise1-Ubuntu SMP Sun Jun 14 18:32:23 UTC 2015 x86_64 x86_64 x86_64 GNU/Linux
- **Install tools**: N/A
- **Others**: N/A

**What happened**:

We have two Kubernetes masters in an HA configuration. During our upgrade, we (of course) upgrade them one at a time. However, after the first master is upgrade, the cluster is in an unstable state.

Namely,
~ Upon taking the first master down, the second master (v1.2) has the key
~ Now we bring the first master up with v1.4, but its master pods fail to start
~ Initially the first master (v1.4) appears as "NodeReady" while the pods attempt to start, but eventually the pods give up, and the first master changes to "NodeNotReady"

We also see this spamming the logs of the second master (v1.2), which retains the key, as the active apiserver:

```
I1006 17:35:54.316991       1 handlers.go:152] PATCH /api/v1/namespaces/default/events/kube-master1.147ac55f24bb7389: (8.736738ms) 200 [[hyperkube/v1.4.0 (linux/amd64) kubernetes/a16c0a7] 192.168.0.9:35588]
I1006 17:35:54.318207       1 handlers.go:152] GET /api/v1/nodes/kube-master1: (4.112012ms) 200 [[hyperkube/v1.4.0 (linux/amd64) kubernetes/a16c0a7] 192.168.0.9:44576]
I1006 17:35:54.324924       1 handlers.go:152] PUT /api/v1/nodes/kube-master1/status: (500.555µs) 415
goroutine 14984574 [running]:
k8s.io/kubernetes/pkg/httplog.(*respLogger).recordStatus(0xc21c6a7500, 0x19f)
    /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/httplog/log.go:214 +0xa6
k8s.io/kubernetes/pkg/httplog.(*respLogger).WriteHeader(0xc21c6a7500, 0x19f)
    /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/httplog/log.go:193 +0x32
k8s.io/kubernetes/pkg/apiserver.(*baseTimeoutWriter).WriteHeader(0xc21b9b7c20, 0x19f)
    /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/apiserver/handlers.go:258 +0xba
k8s.io/kubernetes/pkg/apiserver/metrics.(*responseWriterDelegator).WriteHeader(0xc2173131d0, 0x19f)
    /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/apiserver/metrics/metrics.go:117 +0x53
k8s.io/kubernetes/pkg/apiserver.writeNegotiated(0x7f7c90f6f0e8, 0xc20842af00, 0x0, 0x0, 0x263a900, 0x2, 0x7f7c90dca250, 0xc21f40e200, 0xc2178bedd0, 0x19f, ...)
    /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/apiserver/apiserver.go:429 +0x174
k8s.io/kubernetes/pkg/apiserver.errorNegotiated(0x7f7c7d984958, 0xc21b9b7f80, 0x7f7c90f6f0e8, 0xc20842af00, 0x0, 0x0, 0x263a900, 0x2, 0x7f7c90dca250, 0xc21f40e200, ...)
    /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/apiserver/apiserver.go:441 +0xdd
k8s.io/kubernetes/pkg/apiserver.(*RequestScope).err(0xc2087d0750, 0x7f7c7d984958, 0xc21b9b7f80, 0x7f7c90dca250, 0xc21f40e200, 0xc2178bedd0)
    /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/apiserver/resthandler.go:84 +0x11f
k8s.io/kubernetes/pkg/apiserver.func·034(0xc2173130b0, 0xc21a261200)
    /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/apiserver/resthandler.go:643 +0x4da
k8s.io/kubernetes/pkg/apiserver/metrics.func·001(0xc2173130b0, 0xc21a261200)
    /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/apiserver/metrics/metrics.go:101 +0x269
github.com/emicklei/go-restful.(*Container).dispatch(0xc208422750, 0x7f7c90dca090, 0xc [[hyperkube/v1.4.0 (linux/amd64) kubernetes/a16c0a7] 192.168.0.9:44576]
I1006 17:35:54.327246       1 handlers.go:152] PATCH /api/v1/namespaces/default/events/kube-master1.147ac58cdb8f5184: (8.904674ms) 200 [[hyperkube/v1.4.0 (linux/amd64) kubernetes/a16c0a7] 192.168.0.9:35588]
I1006 17:35:54.350350       1 handlers.go:152] PATCH /api/v1/namespaces/default/events/kube-master1.147ac55f24bb7389: (21.451416ms) 200 [[hyperkube/v1.4.0 (linux/amd64) kubernetes/a16c0a7] 192.168.0.9:35588]
I1006 17:35:54.361433       1 handlers.go:152] PATCH /api/v1/namespaces/default/events/kube-master1.147ac58cdb8f5184: (9.39221ms) 200 [[hyperkube/v1.4.0 (linux/amd64) kubernetes/a16c0a7] 192.168.0.9:35588]
I1006 17:35:54.371348       1 handlers.go:152] PATCH /api/v1/namespaces/default/events/kube-master1.147ac55f24bb7389: (8.417171ms) 200 [[hyperkube/v1.4.0 (linux/amd64) kubernetes/a16c0a7] 192.168.0.9:35588]
I1006 17:35:54.380128       1 handlers.go:152] PATCH /api/v1/namespaces/default/events/kube-master1.147ac58cdb8f5184: (7.57299ms) 200 [[hyperkube/v1.4.0 (linux/amd64) kubernetes/a16c0a7] 192.168.0.9:35588]
```

**What you expected to happen**:

We expect both masters to be in a stable state (thus maintaining HA) after the first master has been upgraded to v1.4 but the second master still remains on v1.2.

**How to reproduce it** (as minimally and precisely as possible):

Have a cluster with two+ masters, with all masters & nodes on v1.2. Upgrade one master to v1.4.

**Anything else do we need to know**:

This appears to be extremely similar to this ticket -- https://github.com/kubernetes/kubernetes/issues/28361 -- however, that ticket involved a single master and was caused by upgrading nodes prior to the master, which is understandably a bad flow.

In our case, we have two HA masters, and need to upgrade them one at a time to maintain HA -- so I believe we need further info on what the root problem is here, and how/whether we can avoid it.

**Comment by bprashanth** [comment]
looks like your upgraded master is not able to write node status to the old master
```
I1006 17:35:54.324924       1 handlers.go:152] PUT /api/v1/nodes/kube-master1/status: (500.555µs) 415
```
to me this sounds like "upgrading nodes before master" (issue stated in https://github.com/kubernetes/kubernetes/issues/28361), or rather, upgrading a client that performs a PUT on node/status before upgrading the thing that's accepting the PUT

**Comment by DreadPirateShawn** [comment]
Right, as we suspected. @bprashanth Are you confirming that this is a bug? Or implying that there's a way to upgrade Kubernetes HA masters one at a time without encountering this problem?

Trying to make sure I understand the next steps here, thanks.

**Comment by fgrzadkowski** [comment]
@jszczepkowski @wojtek-t 

Clarifying questions:
* how do you access apiservers? do you have some load balancer in front of them?
* you mentioned "active apiserver". What do you mean?

The problem that you see with 415 errors is because in 1.3 we are using protocol buffers to communicate between kubelet and apiserver. You updated your master and the kubelet running was trying to register itself and fails because the only running apiserver does not understand proto buffs. This is expected. However it should be unrelated to the reason why your master pods are failing to start. Do you have some logs showing why they failed? In general all the pods running on the master should be able to start regardless of the node registration (this is how it works in single node setup).

**Comment by DreadPirateShawn** [comment]
Apologies for the delay, thought I replied.

We do have a loadbalancer in front of the apiservers, we use haproxy.

My mention of "active apiserver" was probably flawed. I was thinking of the logs from the podmaster on the same master node, which say "key already exists, we are the master" -- but looking again, I realize that only applies to the scheduler and controller manager. The apiserver is separate, and looks like it's active on both masters at once, since it shouldn't matter which of them replies to any given API interaction.

Unfortunately we don't have additional logs at this point, beyond the stack trace from the original description.

(Ultimately we ran into docker/docker#27607 as a blocker issue, and we needed to perform a full rollback. We considered trying to upgrade each separately, but at this point it makes more sense to wait for Docker 1.13 to finalize, and then revisit.)
